DataDog/dd-trace-rb
# Report Linux native thread ids in profiles for Ruby < 3.1

**What does this PR do?**:

This PR adds support for reading the actual native thread id (tid) for Ruby threads in Ruby < 3.1.

Starting in Ruby 3.1, Ruby stores the tid for every thread it creates, and even exposes a `Thread#native_thread_id` with this information.

The tid is relevant because it's what shows up in OS-level tools, such as task managers, or even the Datadog native profiler, thus enabling the correlation of what a thread was doing from both inside and outside the VM.

As mentioned above, up until now, the profiler was only able to report this tid for Ruby 3.1+. For older Rubies, we used a fallback id, which was still as unique as the tid, but was not useful when correlating what a thread was doing with other tools.

This profiler introduces a new `LinuxTidFallback` class, which, together with some dark magic implemented in `linux_tid_from_pthread.c` (see comments on that file for details), is able to retrieve the tid for threads, even though the Linux libc developers really don't want us to do that.

**Motivation**:

Having a tid, as I mentioned above, allows us to correlate what happens inside the Ruby VM with what external tools can observe as well.

The tid is especially visible in the upcoming profiler timeline feature, which is why I decided to finally take a stab at improving this as a R&D week project.

**Additional Notes**:

This new feature is on by default, but can be disabled via configuration.

My intention with making it configurable is as a "just-in-case" lever for support; I do not recommend disabling it.

Although the profiler only officially supports Linux, we like to keep it running on macOS and others for ease of development and experimentation. This feature appropriately degrades on those OSs and disables itself.

There's a few situations, Linux-permissions-wise, where this approach may not work. In those cases (running in CircleCI is one of them), we also degrade gracefully back to the previous behavior.

**How to test the change?**:

This change includes test coverage. You can also see it running by running the profiler on Linux + Ruby < 3.1 and observing that the tids reported are actual linux tids that are seen in any task manager tool such as `htop` (do remember to turn on the feature to see individual threads, since most tools don't show them by default).
